### PR TITLE
Authetication cleanup

### PR DESCRIPTION
**Pull request by hanpeter**
## What does this PR do?

This PR cleans up how authentication token is handled
## Why is this change being made?
1. Simply failing the API call because authentication token has expired is not acceptable.
2. Passing credentials with every method is not ideal.
## Where should the reviewer start?

`catchpoint.py`
## How was this tested? How can the reviewer verify your testing?

The change was manually tested on development environment.

### CODE REVIEW

File: catchpoint.py [hunk 1]
```diff
@@ -27,53 +30,63 @@ def __init__(
...
-        self.content_type = "application/json"
 
         self._logger = logger if logger is not None else getLogger(name=self.__class__.__name__)
-        self._auth = False
-        self._token = None
+        self._headers = {
+            'Accept': 'application/json',
+        }
+        # GOTCHA: Arbitrarily chosen date time in the past
+        self._token_expires_on = datetime(MINYEAR, 1, 1, tzinfo=pytz.utc)
```
Comment (joestump) [inline catchpoint.py:47]: Why is this a gotcha? Can you include a link or more explanation?

File: catchpoint.py [hunk 2]
```diff
@@ -27,53 +30,63 @@ def __init__(
...
-        payload = {
-            'grant_type': 'client_credentials',
-            'client_id': creds['client_id'],
-            'client_secret': creds['client_secret']
-        }
+        now = datetime.now(tz=pytz.utc).replace(microsecond=0)
+        if self._token_expires_on < now:
+            # Remove old authentication header
+            if "Authorization" in self._headers:
+                self._headers.pop("Authorization")
```
Comment (joestump) [inline catchpoint.py]: I feel like `del` is more clear Python for removing a key in a `dict`.

**Review by joestump — APPROVED** [review]
Minor comments, but the code looks good.

File: catchpoint.py [hunk 3]
```diff
@@ -27,53 +30,64 @@ def __init__(
...
-            self._connection_error(e)
+            # Set the new token in header
+            access_token = auth_token["access_token"]
+            self._debug("Access token: " + access_token)
+            self._headers["Authorization"] = "Bearer " + base64.b64encode(access_token)
 
-        self._debug("URL: " + r.url)
-        data = r.json()
+            # Remember the expiry datetime
+            # GOTCHA: Take out 60 seconds from the expiry just to be safe
```
Comment (plathrop) [inline catchpoint.py]: Can you explain this a little better than "just to be safe"? What is the gotcha you experienced here?

File: catchpoint.py [hunk 4]
```diff
@@ -27,53 +30,64 @@ def __init__(
...
+            self._token_expires_on = now + timedelta(seconds=(int(auth_token["expires_in"]) - 60))
```
Comment (plathrop) [inline catchpoint.py]: Prefer constants to hardcoded integers. Something like

``` python
# At class level
EXPIRATION_SAFETY_MARGIN = 60 # seconds

# Here
self._token_expires_on = now + timedelta(seconds=(int(auth_token["expires_in"]) - EXPIRATION_SAFETY_MARGIN))
```

This accomplishes two things: it makes the code more readable to maintainers (I know what this magic number means now), and it makes it easier to change when this value inevitably finds use elsewhere.

File: catchpoint.py [hunk 5]
```diff
@@ -27,53 +30,64 @@ def __init__(
...
+            self._debug("Expires at: " + self._token_expires_on.isoformat())
 
-        self._token = data['access_token']
-        self._debug("TOKEN: " + self._token)
-        self._auth = True
+        return self._headers
```
Comment (plathrop) [inline catchpoint.py:92]: Nit: you called this method `get_header`, but you are returning a _collection_ of headers (plural).

File: catchpoint.py [hunk 6]
```diff
@@ -27,53 +30,64 @@ def __init__(
...
 
-    def _make_request(self, uri, params=None, data=None):
+    def _call(self, url, *args, **kwargs):
```
Comment (plathrop) [inline catchpoint.py:94]: Whenever you find yourself accepting `*args, **kwargs` you know immediately that you _must_ provide a detailed comment explaining why it was necessary.

File: catchpoint.py [hunk 7]
```diff
@@ -82,13 +96,17 @@ def _make_request(self, uri, params=None, data=None):
...
-        res = requests.get(final_url, headers=headers, params=params, data=data)
+        final_url = self._URL_TEMPLATE.format(host=self._host, version=self._version, uri=url)
+        return self._make_request(
+            url=final_url,
+            headers=self._get_header(),
+            *args,
+            **kwargs
+        )
+
+    def _make_request(self, *args, **kwargs):
```
Comment (plathrop) [inline catchpoint.py:113]: Ditto re: `*args, **kwargs`. Also, docstring (yes, even internal methods deserve docstrings).

**Review by plathrop — CHANGES_REQUESTED** [review]
Looks good overall, but there are a couple cleanup items necessary.

**Comment by hanpeter** [comment]
@plathrop All your comments should be addressed. There needs to be a full update on docstring and `README.md`, but I didn't want to blow up the scope of this PR. I'll handle that in the next PR.

**Review by plathrop — APPROVED** [review]
👍

**Pull request merged**